transports/tcp/src/lib.rs

```diff
@@ -441,9 +505,106 @@ where
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        // We remove each element from `listeners` one by one and add them back.
+        let mut remaining = self.listeners.len();
+        while let Some(mut listener) = self.listeners.pop_back() {
+            match TryStream::try_poll_next(listener.as_mut(), cx) {
+                Poll::Pending => {
+                    self.listeners.push_front(listener);
+                    remaining -= 1;
+                    if remaining == 0 {
+                        break;
+                    }
+                }
+                Poll::Ready(Some(Ok(TcpListenerEvent::Upgrade {
+                    upgrade,
+                    local_addr,
+                    remote_addr,
+                }))) => {
+                    let id = listener.listener_id;
+                    self.listeners.push_front(listener);
+                    return Poll::Ready(TransportEvent::Incoming {
+                        listener_id: id,
+                        upgrade,
+                        local_addr,
+                        send_back_addr: remote_addr,
+                    });
+                }
+                Poll::Ready(Some(Ok(TcpListenerEvent::NewAddress(a)))) => {
+                    let id = listener.listener_id;
+                    self.listeners.push_front(listener);
+                    return Poll::Ready(TransportEvent::NewAddress {
+                        listener_id: id,
+                        listen_addr: a,
+                    });
+                }
+                Poll::Ready(Some(Ok(TcpListenerEvent::AddressExpired(a)))) => {
+                    let id = listener.listener_id;
+                    self.listeners.push_front(listener);
+                    return Poll::Ready(TransportEvent::AddressExpired {
+                        listener_id: id,
+                        listen_addr: a,
+                    });
+                }
+                Poll::Ready(Some(Ok(TcpListenerEvent::Error(error)))) => {
+                    let id = listener.listener_id;
+                    self.listeners.push_front(listener);
+                    return Poll::Ready(TransportEvent::ListenerError {
+                        listener_id: id,
+                        error,
+                    });
+                }
+                Poll::Ready(None) => {
+                    return Poll::Ready(TransportEvent::ListenerClosed {
+                        listener_id: listener.listener_id,
+                        reason: Ok(()),
+                    });
+                }
+                Poll::Ready(Some(Err(err))) => {
+                    return Poll::Ready(TransportEvent::ListenerClosed {
+                        listener_id: listener.listener_id,
+                        reason: Err(err),
+                    });
+                }
+            }
+        }
+        // We register the current task to be woken up if a new listener is added.
```

Comment:
Where is the waker of the current task stored to be woken up later?

Reply:
I copied that logic, including the comment, from the old `swarm::connection::ListenersStream::poll`. 
It seems like a left-over from before we even used stable futures and probably only served to explain how `Poll::Pending` (back then `Ok(Async::NotReady)`) works.
Remove it now. As we also discussed it in https://github.com/libp2p/rust-libp2p/pull/2652#discussion_r906871039. we assume that the transport is polled again after a new listener is added, which will then register the new listeners waker.